### ISSUE TITLE
how to make Ads in midroll using clappr player, can someone advise?

### ISSUE THREAD

**Issue by kaushik-hemant** [clappr/clappr#1244]
**Browser**: FILL WITH YOUR BROWSER (ex: Chrome Version 46.0.2490.80, Firefox Version 3.5.6, IE 11)

**OS**: FILL WITH YOUR OS (ex: Mac OS 10.11.1, iOS9, android4.5)

**Clappr Version**: 0.2.25

**Steps to reproduce**:

* first step
* then second step
* I was expecting X but instead it shows Y

Did you try to reproduce this issue at http://cdn.clappr.io/

ps: you can attach images, logs or whatever you think might be helpful.
ps: feel free to clean this if your issue is different.

**Comment by kslimani** [comment]
Technically, it should be almost the same as preroll, except for VOD source : you need to store content video current position and restore it when playing back.

**Comment by kaushik-hemant** [comment]
Can you please share some code for example for pre post and mid?

Sent from my iPhone

On 27-Dec-2016, at 7:09 PM, kslimani <notifications@github.com<mailto:notifications@github.com>> wrote:


Technically, it should be almost the same as preroll, except for VOD source : you need to store content video current position and restore it when playing back.

**Comment by kslimani** [comment]
I don't have example code for midroll and postroll, but you may find something usefull here :

* https://github.com/vix-simplex/clappr-vast-ad-plugin
* https://github.com/mycujoo/clappr-vast-plugin
* https://github.com/kslimani/clappr-google-ima-html5-preroll
* https://github.com/kslimani/clappr-html5-preroll-skeleton-plugin

**Comment by kaushik-hemant** [comment]
I surf around for midroll but didnt found anything useful, can i create my own plugin on top of clapper with time slot like at 00:10 schedule a ad or so.

Sent from my iPhone

On 27-Dec-2016, at 7:28 PM, kslimani <notifications@github.com<mailto:notifications@github.com>> wrote:


I don't have example code for midroll and postroll, but you may find something usefull here :

  *   https://github.com/vix-simplex/clappr-vast-ad-plugin
  *   https://github.com/mycujoo/clappr-vast-plugin
  *   https://github.com/kslimani/clappr-google-ima-html5-preroll
  *   https://github.com/kslimani/clappr-html5-preroll-skeleton-plugin

-

**Comment by kslimani** [comment]
Of course you can create your own plugin. Take a look at [Clappr Architecture](https://github.com/clappr/clappr/wiki/Architecture) and at previously provided links for plugin base examples.

**Comment by leandromoreira** [comment]
One thing you can try is to create two playbacks then you can play with the main `[0]` and listen to its `timeupdate` and when it reaches `X` time you pause it and switch the playbacks order in the container.

**Comment by kslimani** [comment]
@leandromoreira i am not sure this trick will work onmobile device because of user interaction requirement to start video. _(in HTML5 video element scenario that is)_. But yes, it worth a try.

**Comment by leandromoreira** [comment]
@kslimani yes, you're right, I suspect @towerz or @bernardocamilo already done something similar to this...

**Comment by kslimani** [comment]
@leandromoreira i think @towerz is aware of this issue because i remember reading something where he mention about recycle the video element when instancing another playback on loading another source in core.

**Comment by leandromoreira** [comment]
Exactly, this is already an issue today, given to the scenario where we have to deliver Ads on pre-roll that uses the same videotag

**Comment by kslimani** [comment]
Yes, and the "recyle" mention is in the [projet dashboard](https://github.com/clappr/clappr/projects/1) _(the fourth ticket in TODO list)_. So i think the plan is to fix it in next major release.

**Comment by thalleshmm** [comment]
I just published a plugin for pre-roll and post-roll. It still needs some work, but it does the job. I hope it helps.

https://github.com/JMVTechnology/Clappr-Ads
